tests/test_integration.py

```diff
@@ -82,150 +109,166 @@ def _create_vm():
-migratevm = partial(real_migratevm, balance_config='test')
+    return cmd.format(*escaped_args, **escaped_kwargs)
 
 
 class IGVMTest(unittest.TestCase):
-    def _check_vm(self, hv, vm):
-        fqdn = vm.run('hostname -f').strip()
-        self.assertEqual(fqdn, vm.fqdn)
-
-        self.assertEqual(vm.server_obj.is_dirty(), False)
-
-        self.assertEqual(hv.vm_defined(vm), True)
-        self.assertEqual(hv.vm_running(vm), True)
-
-    def _check_absent(self, hv, vm):
-        self.assertEqual(hv.vm_defined(vm), False)
-        hv.run('test ! -b /dev/xen-data/{}'.format(vm.fqdn))
-
-
-class BalanceBuildTest(IGVMTest):
-    """ Same as BuildTest but with tearDown and setUp cleaning *both* HVs """
-
     @classmethod
     def setUpClass(cls):
-        cls.hv1 = Hypervisor(HV1)
-        cls.hv2 = Hypervisor(HV2)
-        cls.vm = _create_vm()
+        cls.hvs = HYPERVISORS
 
-        _clean_vm((cls.hv1, cls.hv2), cls.vm)
+    def setUp(self):
+        """Initialize VM object before every test.
+
+        Get object from Serveradmin and initialize it to safe defaults.
+        Don't assign VM to any of HVs yet!
+        """
+
+        # igvm operates always on hostname of VM and queries it from
+        # Serveradmin whenever it needs. Because of that we must never store
+        # any igvm objects and query things anew each time.
+        self.vm_obj = self.get_vm_obj()
+
+        # Fill in defaults in Serveradmin
+        self.vm_obj['state'] = 'online'
+        self.vm_obj['disk_size_gib'] = 3
+        self.vm_obj['memory'] = 2048
+        self.vm_obj['num_cpu'] = 2
+        self.vm_obj['os'] = 'jessie'
+        self.vm_obj['environment'] = 'testing'
+        self.vm_obj['no_monitoring'] = True
+        self.vm_obj['xen_host'] = None
+        self.vm_obj['repositories'] = [
+            'int:basejessie:stable',
+            'int:innogames:stable jessie',
+        ]
+        self.vm_obj['puppet_environment'] = None
+        self.vm_obj.commit()
 
-    @classmethod
-    def tearDownClass(cls):
-        _clean_vm((cls.hv1, cls.hv2), cls.vm)
+    def tearDown(self):
+        """Clean up all HVs after every test."""
+
+        for hv in self.hvs:
+            hv.run(
+                'virsh destroy {vm}; '
+                'virsh undefine {vm}'
+                .format(vm=self.vm_obj['hostname']),
+                warn_only=True,
+            )
+            hv.run(
+                'umount /dev/xen-data/{vm}; '
+                'lvremove -f /dev/xen-data/{vm}'
+                .format(vm=self.vm_obj['hostname']),
+                warn_only=True,
+            )
 
-    def test_buildvm_auto_find_hypervisor(self):
-        self.vm.server_obj['xen_host'] = None
-        self.vm.server_obj.commit()
+    def get_vm_obj(self):
+        vm_obj = Query(
+            {
+                'servertype': 'vm',
+                'hostname': VM_HOSTNAME,
+            },
+        )
+        return vm_obj.get()
```

Comment:
Then we need real code in quite many places. Isn't this what functions are for, to not repeat real code?

Reply:
I just don't see the gain here, as the full query is already very brief. It's one less indirection I have to resolve in my head and it would probably still be less lines of code, but YMMV.